core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighter.java

```diff
@@ -106,7 +107,12 @@ public HighlightField highlight(HighlighterContext highlighterContext) {
             textsToHighlight = HighlightUtils.loadFieldValues(field, mapper, context, hitContext);
 
             for (Object textToHighlight : textsToHighlight) {
-                String text = textToHighlight.toString();
+                String text;
+                if (textToHighlight instanceof BytesRef) {
+                    text = mapper.fieldType().valueForDisplay(textToHighlight).toString();
+                } else {
+                    text = textToHighlight.toString();
+                }
```

Comment:
The type of textToHighlight can be different whether it comes from the _source or from the stored fields. Calling valueForDisplay on it when it comes from _source would not work.